observatory_platform/telescopes/crossref_events.py

```diff
@@ -0,0 +1,701 @@
+    """
+    Transforms the crossref events release. The download file contains multiple lists, one for each request,
+    and each list contains multiple events. Each event is transformed so that the field names do not contain
+    '-' and have a valid timestamp at 'occurred_at'. The events are written out individually and separated by a newline.
+
+    :param release: The crossref events release
+    :return: None
+    """
+    # load events
+    events = []
+    with open(release.download_path, 'r') as f:
+        for line in f:
+            events += json.loads(line)
+
+    # transform release
+    with open(release.transform_path, 'w') as f:
+        for event in events:
+            try:
+                pendulum.parse(event['occurred_at'])
+            except ValueError:
+                event['occurred_at'] = "0001-01-01T00:00:00Z"
+            event_updated = change_keys(event, convert)
+            f.write(json.dumps(event_updated) + '\n')
+
+
+def convert(k: str) -> str:
+    """
+    Replaces '-' with '_'
+
+    :param k: Dictionary key
+    :return: Updated dictionary key
+    """
+    return k.replace('-', '_')
+
+
+def change_keys(obj, convert):
+    """
+    Recursively goes through the dictionary obj and updates keys with the convert function.
+
+    :param obj: The dictionary
+    :param convert: The convert function that is used to update the key
+    :return: The updated dictionary
+    """
+    if isinstance(obj, (str, int, float)):
+        return obj
+    if isinstance(obj, dict):
+        new = obj.__class__()
+        for k, v in obj.items():
+            new[convert(k)] = change_keys(v, convert)
+    elif isinstance(obj, (list, set, tuple)):
+        new = obj.__class__(change_keys(v, convert) for v in obj)
+    else:
+        return obj
+    return new
+
+
+class CrossrefEventsRelease:
+    """ Used to store info on a given crossref events release """
+
+    def __init__(self, start_date: pendulum.Pendulum, end_date: pendulum.Pendulum, first_release: bool = False):
+        """
+        :param start_date: Start date of this release
+        :param end_date: End date of this release
+        :param first_release: Whether this is the first release that is downloaded (if so, no edited/deleted events
+        need to be obtained)
+        """
+        self.start_date = start_date
+        self.end_date = end_date
+        self.first_release = first_release
+
+        if CrossrefEventsTelescope.DOWNLOAD_MODE == 'parallel':
+            batches = self.batch_dates
+        elif CrossrefEventsTelescope.DOWNLOAD_MODE == 'sequential':
+            batches = [(start_date, end_date)]
+        else:
+            raise AirflowException(f'Download mode has to be either "sequential" or "parallel", '
+                                   f'not "{CrossrefEventsTelescope.DOWNLOAD_MODE}"')
+
+        self.urls = []
+        edited_deleted = False
+        for batch in batches:
+            start_date = batch[0]
```

Comment:
Not sure if you want to use this or not, but if you do, in the scopus branch there's a new class defined called SchedulePeriod that has .start and .end attributes of type pendulum.date for a little more human readability.

Reply:
Thanks, I'll have a look!